game/gamestate.go

```diff
@@ -40,13 +55,154 @@ func (gs *GameState) Copy() *GameState {
 	ownershipCopy := make([]int, len(gs.Ownership))
 	copy(ownershipCopy, gs.Ownership)
 
+	// Deep copy PlayerTroops
+	playerTroopsCopy := make(map[int]int)
+	for key, value := range gs.PlayerTroops {
+		playerTroopsCopy[key] = value
+	}
+
 	return &GameState{
-		Map:         gs.Map,
-		TroopCounts: troopCountsCopy,
-		Ownership:   ownershipCopy,
+		Map:           gs.Map,
+		TroopCounts:   troopCountsCopy,
+		Ownership:     ownershipCopy,
+		Rules:         gs.Rules,
+		CurrentPlayer: gs.CurrentPlayer,
+		LastMove:      gs.LastMove,
+		Phase:         gs.Phase,
+		PlayerTroops:  playerTroopsCopy,
+		TroopsToPlace: gs.TroopsToPlace,
 	}
 }
 
+// LegalMoves returns all legal moves for the current player.
+func (gs *GameState) LegalMoves() []Move {
+	switch gs.Phase {
+	case ReinforcementPhase:
+		return gs.reinforcementMoves()
+	case AttackPhase:
+		return gs.attackMoves()
+	case ManeuverPhase:
+		return gs.maneuverMoves()
+	default:
+		return nil
+	}
+}
+
+// reinforcementMoves generates all possible reinforcement moves for the current player.
+func (gs *GameState) reinforcementMoves() []Move {
+	var moves []Move
+	remainingTroops := gs.TroopsToPlace
+
+	// Get enemy-adjacent territories
+	enemyAdjacentTerritories := gs.getEnemyAdjacentTerritories()
+
+	// Possible troop amounts: one, half, all
+	troopAmounts := []int{1, remainingTroops / 2, remainingTroops}
+
+	for _, cantonID := range enemyAdjacentTerritories {
+		for _, amount := range troopAmounts {
+			if amount > 0 && amount <= remainingTroops {
+				moves = append(moves, &GameMove{
+					ActionType: ReinforceAction,
+					ToCantonID: cantonID,
+					NumTroops:  amount,
+				})
+			}
+		}
+	}
+	return moves
+}
+
+func (gs *GameState) attackMoves() []Move {
+	var moves []Move
+
+	for cantonID, owner := range gs.Ownership {
+		if owner == gs.CurrentPlayer && gs.TroopCounts[cantonID] > 1 {

```

Comment:
can use `gs.getEnemyAdjacentTerritories()` here to stay consistent with reinforcement logic